### PR TITLE
[REFERENT SPACE] Save referent space page access dates

### PR DESCRIPTION
**Pull request by illusionOfParadise**

### CODE REVIEW

File: src/EventListener/ReferentSpacePageVisitedListener.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    public function onKernelResponse(FilterResponseEvent $event)
+    {
+        $request = $event->getRequest();
+
+        // Stop for non referent space routes
+        if (0 !== mb_strpos($request->getRequestUri(), '/espace-referent')) {
+            return;
+        }
+
+        // Stop if no Session or it's not an referent
```
Comment (JeremieSamson) [inline src/EventListener/ReferentSpacePageVisitedListener.php]:
```suggestion
        // Stop if no Session or if it's not a referent
```

File: src/Repository/ReferentSpaceAccessInformationRepository.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+class ReferentSpaceAccessInformationRepository extends ServiceEntityRepository
+{
+    public function __construct(RegistryInterface $registry)
+    {
+        parent::__construct($registry, ReferentSpaceAccessInformation::class);
+    }
+
+    public function findByAdherent(Adherent $adherent)
+    {
+        return $this->findOneBy(['adherent' => $adherent]);
```
Comment (ottaviano) [inline src/Repository/ReferentSpaceAccessInformationRepository.php]: just one idea, what about:
```php
function findByAdherent(Adherent $adherent): ReferentSpaceAccessInformation {
  $obj = $this->findOneBy(['adherent' => $adherent]);
  if (!$obj) {
    $obj = new ReferentSpaceAccessInformation($adherent, new \Date, new \Date);
  }

  return $obj;
}
```

File: tests/Entity/ReferentSpaceAccessInformationTest.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+namespace Tests\AppBundle\Entity;
+
+use AppBundle\DataFixtures\ORM\LoadAdherentData;
+use AppBundle\Entity\ReferentSpaceAccessInformation;
+use Liip\FunctionalTestBundle\Test\WebTestCase;
+use Tests\AppBundle\TestHelperTrait;
+
+class ReferentSpaceAccessInformationTest extends WebTestCase
+{
+    use TestHelperTrait;
```
[the remark this answers is not in the document]
Comment (illusionOfParadise) [reply tests/Entity/ReferentSpaceAccessInformationTest.php]: It's used here https://github.com/EnMarche/en-marche.fr/pull/3554/files#diff-1c07008dae738b7ef902862d1e3c1474R39

File: tests/Entity/ReferentSpaceAccessInformationTest.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    private function createReferentSpaceAccessInformation(
+        string $uuid,
+        string $datetime
+    ): ReferentSpaceAccessInformation {
+        $adherent = $this->getAdherent($uuid);
+
+        return new ReferentSpaceAccessInformation($adherent, new \DateTimeImmutable($datetime), new \DateTimeImmutable($datetime));
+    }
+
+    protected function setUp()
```
[the remark this answers is not in the document]
Comment (illusionOfParadise) [reply tests/Entity/ReferentSpaceAccessInformationTest.php]: Needed for `TestHelperTrait`.

File: src/Repository/ReferentSpaceAccessInformationRepository.php (hunk 2 above)
Comment (illusionOfParadise) [reply src/Repository/ReferentSpaceAccessInformationRepository.php]: Seems strange doing this in a `find` method.
That is cheating😄

File: tests/Entity/ReferentSpaceAccessInformationTest.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (illusionOfParadise) [reply tests/Entity/ReferentSpaceAccessInformationTest.php]: I have changed the test. It's no more used.

File: src/Entity/ReferentSpaceAccessInformation.php [hunk 5]
```diff
@@ -0,0 +1,96 @@
+<?php
+
+namespace AppBundle\Entity;
+
+use Algolia\AlgoliaSearchBundle\Mapping\Annotation as Algolia;
+use Doctrine\ORM\Mapping as ORM;
+
+/**
+ * @ORM\Table(name="referent_space_access_information")
```
Comment (ottaviano) [inline src/Entity/ReferentSpaceAccessInformation.php]: `name` is useless here

File: src/EventListener/RecordReferentLastVisitListener.php [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
+use Symfony\Component\Security\Core\Security;
+
+class RecordReferentLastVisitListener implements EventSubscriberInterface
+{
+    private $security;
+    private $authorizationChecker;
+    private $manager;
+    private $referentSpaceAccessInformationRepository;
```
Comment (ottaviano) [inline src/EventListener/RecordReferentLastVisitListener.php]: what about just `$repository`? it's really long as a name :P

File: src/Repository/ReferentSpaceAccessInformationRepository.php [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+use Symfony\Bridge\Doctrine\RegistryInterface;
+
+class ReferentSpaceAccessInformationRepository extends ServiceEntityRepository
+{
+    public function __construct(RegistryInterface $registry)
+    {
+        parent::__construct($registry, ReferentSpaceAccessInformation::class);
+    }
+
+    public function findByAdherent(Adherent $adherent)
```
Comment (ottaviano) [inline src/Repository/ReferentSpaceAccessInformationRepository.php]: typehint pls

File: src/Entity/ReferentSpaceAccessInformation.php (hunk 5 above)
Comment (illusionOfParadise) [reply src/Entity/ReferentSpaceAccessInformation.php]: No, it's in the singular by default.

**Pull request merged**